### DISCUSSION TITLE
1. Rich Type Annotations

### DISCUSSION THREAD

**Discussion by pysan3** [nvim-neo-tree/neo-tree.nvim#1335] [Internal Development]
Hi y'all, 

I think I can start working on the big rewrite.
I am going to bring a huge diff to the codebase, so I'd like to ask your opinion first.

I will post my roadmaps here, separated by its topics.
I'd love to hear your feedbacks on each of the threads.

- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1335
- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1336
- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1337
- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1338
- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1339

### 1. Rich Type Annotations

I see some type annotations added to the codebase recently, which is amazing.

I'd like to scan thru the codebase to define the fields for the `context` and `state` variable, which is used and passed everywhere, and many fields are added along the way. (Use `---@class, ---@field`.)
Also, add annotation and doc for the `default.lua` to give a better configuration experience. I see some complaints that neo-tree's config options are too complicated.

### 2. Add Type Check Action

Ref: https://github.com/mrcjkb/lua-typecheck-action/issues/34

After `1.` is done, we can check for type errors with GH actions for each PR, which makes it way easier to detect potential bugs.

I started working: https://github.com/nvim-neo-tree/neo-tree.nvim/pull/1333
I also added `lua_ls-typecheck.yml`, which is configured to only check types on `defaults.lua` for now, and I will gradually make the list longer as I add the correct annotations.

cc @cseickel

**Comment by cseickel** [comment]
This one is very simple: types are great! The only reason it does not have type already is that they either didn't exist or I didn't know about them when I started the project.

**Reply by pysan3** [reply]
Thanks, I'll first make everything type correct before any other tasks ;)

BTW, who else do you think should I ping here?

**Reply by cseickel** [reply]
@miversen33 AND @nhat-vo are the other maintainers that should be aware of this series of discussions. @ghostbuster91 is also writing an external source and becoming an active contributor, so he should see this series too.

https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1336
https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1337
https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1338
https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1339
